### PR TITLE
Allow different splitting rules

### PR DESCRIPTION
**Pull request by velochy**
This PR is inspired by the wish to better handle categorical predictors, and basically implements the paper https://arxiv.org/abs/2211.04459 

Not sure how the PR process works for your repo, but happy to comply with any and all standards to get the changes merged :)

### CODE REVIEW

**Comment by aloctavodia** [comment]
Thanks for this PR!

Not sure we have a contributing guide for this repo, or it may be outdated (sorry about that). But in order for a PR to be accepted, the code should be formatted with `black -l 100`, pylint should pass, ideally we should have type hints (we use mypy). And tests should be added.

**Comment by aloctavodia** [comment]
@juanitorduz as you have some experience with BART and categorical variables you may have comments about this.

File: pymc_bart/split_rules.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+    @staticmethod
+    @njit
+    def divide(available_splitting_values, split_value):
+        return available_splitting_values == split_value
+
+class SubsetSplitRule:
+    """
+    Choose a random subset of the categorical values and branch on if the value is within the chosen set.
+    This is the approach taken in flexBART paper.
```
Comment (aloctavodia) [inline pymc_bart/split_rules.py:55]: Please add a reference to the paper including name of the article, authors and DOI.

File: pymc_bart/__init__.py [hunk 2]
```diff
@@ -15,6 +15,7 @@
 
 from pymc_bart.bart import BART
 from pymc_bart.pgbart import PGBART
+from pymc_bart.split_rules import *
```
Comment (juanitorduz) [inline pymc_bart/__init__.py:18]: Shall we avoid `import *` statements?

File: pymc_bart/bart.py [hunk 3]
```diff
@@ -109,6 +109,7 @@ def __new__(
         beta: float = 2.0,
         response: str = "constant",
         split_prior: Optional[List[float]] = None,
+        split_rules: Optional[List] = None,
```
Comment (juanitorduz) [inline pymc_bart/bart.py:112]: Shall we keep the exact expected types (maybe `ContinuousSplitRule` from what I see below)?

**Review by juanitorduz — COMMENTED** [review]
This looks very interesting! Whenever I need to work with categorical variables, I actively omit the one-hot-encoding (reasons similar to the ones presented in the paper), and rather use variations of target encoding (see [Category Encoders](https://contrib.scikit-learn.org/category_encoders/)).

The idea behind `flexBART` seems very interesting, especially the "pooling" between categories through the tree structures. I do not know the full details, and I think @aloctavodia is the expert to review the code in detail. I left some small comments regarding the code. I will be happy to help regarding type-hints and looking into tests :)

File: pymc_bart/split_rules.py (hunk 1 above)
Comment (velochy) [reply pymc_bart/split_rules.py:55]: It's this paper: https://arxiv.org/abs/2211.04459
Do you have an example of citation formating somewhere so I can align the style?

File: pymc_bart/__init__.py (hunk 2 above)
Comment (velochy) [reply pymc_bart/__init__.py:18]: Will fix that :)

File: pymc_bart/bart.py (hunk 3 above)
Comment (velochy) [reply pymc_bart/bart.py:112]: This is where I need a bit of help. If you look at split_rules.py you see there are multiple different split rule options.
They essentially form an abstract class, but as they have staticmethods (for performance reasons) I don't know if it is possible to abstract them into a SplitRule class type in any reasonable way that actually enforces the static methods existing?

Comment (velochy) [reply pymc_bart/bart.py]: Did my best here: created a class SplitRule with two static abstract methods and made other split rules inherit from it, then added SplitRule as hint here.

Comment (juanitorduz) [reply pymc_bart/bart.py]: ok! It seems `mypy` does not like abstract classes type=hints (https://github.com/python/mypy/issues/4717. My suggestion is to focus on the core functionalities and test and come back to this minor detail once we are ready to merge :) (From my side, I will investigate how to do it)

File: pymc_bart/split_rules.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
+#   Copyright 2022 The PyMC Developers
```
[the remark this answers is not in the document]
Comment (velochy) [reply pymc_bart/split_rules.py:1]: I can try. Don't have much experience with pytest but I can hopefully figure it out :)

[the remark this answers is not in the document]
Comment (aloctavodia) [reply pymc_bart/split_rules.py:1]: The tests in https://github.com/pymc-devs/pymc-bart/blob/main/tests/test_tree.py are probably the simplest ones in PyMC-BART, you can try following a similar pattern.

[the remark this answers is not in the document]
Comment (velochy) [reply pymc_bart/split_rules.py:1]: Added basic functionality tests to https://github.com/velochy/pymc-bart/blob/main/tests/test_split_rules.py

**Comment by velochy** [comment]
I have now addressed all the comments to the best of my knowledge.
@juanitorduz do you have any more comments or improvement recommendations?

**Comment by aloctavodia** [comment]
You need to run `black -l 100 pymc_bart/tree.py`

**Comment by aloctavodia** [comment]
Thanks @velochy!

**Pull request merged**